### ISSUE TITLE
Java Tokenize '>>' and '>' '>'

### ISSUE THREAD

**Issue by revintec** [antlr/antlr4#1223]
Hi.

In Java8, there are two tokens '>>'(shift right) and '>''>'(two '>'s).
If I define '>>' as SHIFT_RIGHT, then I can't have rule: genericType: type'<'type'>'.
Consider this:
1 >> 2 // '>>' is matched correctly
TypeA&lt;TypeB&lt;TypeC,TypeD>> // which matches a '>>' but really should be seen as two '>'s

After reading Java8.g4, I found the following line:
shiftExpression
    :   additiveExpression
    |   shiftExpression '<' '<' additiveExpression
    |   shiftExpression '>' '>' additiveExpression
    |   shiftExpression '>' '>' '>' additiveExpression
    ;
which is NOT CORRECT because it allows things like:
a > > b // which javac reports error while expecting a >> b;

There seems no way around it, if I have '>>' defined, ANTLR4 won't try to re-interpret '>>' as two '>'s even if not doing so would cause overall grammar match to fail.

Can you guys upgrade ANTLR4 to support lexer backtracking?
Basically, if overall match is a failure, try to use smaller lengthed tokens.
I'm new to parsing and all, but I found C++11 and Java8 all supports this,
wondering if it's very hard to implement this in ANTLR4?

I do understand lexing stage will only return the longest match(which is '>>'),
but why can't we break it down into two '>'s in parsing stage? 

The following way can work around this problem but are very ugly:
1) define SHIFT_RIGHT as '>''>' and check if any blank is present between '>'s in code.
2) define generic as type'<'type'>'|type'<'type'<'type'>>'

**Comment by parrt** [comment]
Yes, a semantic predicate really needs to check that the columns are next to each other. lexer backtracking will not help you here.  only context can tell you what you can expect.

**Comment by sharwell** [comment]
:memo: We don't even need a semantic predicate here. The whitespace check should appear in a listener which runs after the parse tree is created and reports errors if the `<` or `>` characters are not in sequence.
